Website query API: Continue plugin installs on error

eliot-akira opened the pull request:
## What is this PR doing?

Add a callback `onBlueprintStepError` to gather errors, so that plugin installs via website query parameter API can complete even if any fails. These errors should be displayed in a notification.

## What problem is it solving?

Issues WordPress/wordpress-playground#600 and WordPress/wordpress-playground#604, both caused due to entire blueprint steps getting stopped when a plugin fails to install.

## How is the problem addressed?

TBD

## Testing Instructions

TBD

eliot-akira commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -127,9 +131,20 @@ export function compileBlueprint(
-				for (const { run, step } of compiled) {
-					const result = await run(playground);
-					onStepCompleted(result, step);
+				if (onStepError) {
+					for (const { run, step } of compiled) {
+						try {
+							const result = await run(playground);
+							onStepCompleted(result, step);
+						} catch (error) {
+							onStepError(error, step);
+						}
+					}
+				} else {
+					for (const { run, step } of compiled) {
+						const result = await run(playground);
+						onStepCompleted(result, step);
+					}
```
At first I defined a default `onStepError` callback to re-throw the error, but I removed it to maintain current behavior (no try/catch for each step) if the callback is not passed as an option. That way, the error stack trace is cleaner for all other use cases other than the website query API.

eliot-akira commented on packages/playground/website/src/lib/hooks.ts:
```diff
@@ -36,10 +40,22 @@ export function usePlayground({ blueprint, storage }: UsePlaygroundOptions) {
+			step: StepDefinition
+		) => {
+			if (step.step === 'installPlugin') {
+				// Gather errors and show in notification
```
It's a question what to do here with the gathered errors, before we implement the toast notification.

dmsnell commented:
so fast @eliot-akira! thanks - I might need a few days or more to review this.

eliot-akira commented:
Oops, when I added you as reviewer, it said "Request review" but this isn't ready for review yet. It's just a rough draft to get started, I haven't checked if this solves the issue.

I'm making some comments to think about how to best implement the solution, and will develop it further. I still have some doubts, so any feedback is welcome. When it's more in shape I'll let you know by switching the PR from Draft status to "Ready for Review".

eliot-akira commented:
OK, the basic idea of this PR is ready. I confirmed that it does solve the issue and achieve the goal of letting plugin installs (and the rest of blueprint steps) continue even if one fails.

To explain the proposed solution: in the function `compileBlueprint`, it adds an optional callback `onStepError` ([here](https://github.com/WordPress/wordpress-playground/pull/605/files#diff-111618e10ba828ef4596cfba7f3a377a0c7cefc4b2432d96a13a9bd77f462987R59-R60)) which when defined will be called with any error that happens while running the blueprint step. The callback can decide what to do with the error, such as gathering them for notification, or rethrowing it to stop running the rest of steps.

In the Playground website, when blueprint steps are compiled from the URL query API, it defines this error callback ([here](https://github.com/WordPress/wordpress-playground/pull/605/files#diff-24d43f4099f5d879119052735ed9b206bb6c48b68c8bf377c50dfded306b519eR43)). For now, if the error happens in the `installPlugin` step, it just outputs to console error. For any other steps, it rethrows the error.

I tried to implement this in a way that maintains current behavior as much as possible, except for where it's needed in the Query API, so that errors are thrown as expected for any other use of Blueprints (such as via Playground Client module, or on server side). I also tried to make it generally useful, in case there are other kinds of steps that allowed to fail without stopping the entire process.

If you agree with this approach, I can add a test for the `onStepError` option.

dmsnell commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -118,18 +124,24 @@ export function compileBlueprint(
 						resource.setPlayground(playground);
 						if (resource.isAsync) {
-							resource.resolve();
+							resource.resolve().catch(function (error) {
```
is there an existing bug here where we should be awaiting the response from `.resolve()`? It looks like we may continue to lines 130-144 below before all the resources have resolved.

this may be correct, as we want to queue up all the async resources to start downloading in parallel, but we don't wait for the batch of them to complete.

I guess in `resolveArguments` we end up explicitly awaiting the results so that comes through. hm. this could use a comment at a minimum.

it seems that the `onStepError` callback might execute in the middle of processing or waiting for other resources. that also seems like it's probably fine, but possibly surprising.

in other words, this doesn't particularly give us an opportunity to choose to abort initializing the playground if a step fails even though it gives us the chance to know _while_ initializing if one or more things fail.

---

as someone using the Playground I wonder what I care about:
 - I don't care if one plugin fails to load as long as I can know it failed. keep going.
 - if my one particular thing fails then there's no point continuing.

in the end, I guess there's no harm in doing a partial boot of the playground as long as we're not connected to an OPFS or Live Directory share. partial boot in those places could corrupt files, though I hope people aren't relying on this _not_ corrupting files. it's supposed to be ephemeral.

---

did you consider passing a callback when we finish running the steps?

the current callbacks speak to individual steps, but not to the overall progress.

```ts
const progressObject = new Map<StepDefinition, Promise<void>>();
const progressPromises = new Map<StepDefinition, {resolve: () => void, reject: (error: string) => void}>();
for (const { step } of compiled) {
	progressObject.set(step, new Promise((resolve, reject) => {
		progressPromises.set(step, {resolve, reject});
	} ) );
}
onRunningSteps( progressObject );
```

then when resolving we can do this…
```ts
	for (const { run, step } of compiled) {
		try {
			const result = await run(playground);
			progressPromises.get(step).resolve();
		} catch (error) {
			progressPromises.get(step).reject();
		}
	}
} finally {
	try {
		if (onStepsCompleted) {
			await onStepsCompleted( steps );
		}
		await (playground as any).goTo(
			blueprint.landingPage || '/'
		);
	} catch (e) {
		/*
		 * NodePHP exposes no goTo method.
		 * We can't use `goto` in playground here,
		 * because it may be a Comlink proxy object
		 * with no such method.
		 */
	}
	progress.finish();
}
```

Implementors could still replicate the behavior that `onStepCompleted` and `onStepError` provide, but they could additional prevent the boot-up of the playground.

```ts
// Prevent booting if anything fails.
onStepsCompleted = async ( steps ) => {
	const results = await Promise.allSettled( steps.values() );
	if ( results.any( p => p.status === 'rejected' ) ) {
		throw new Error( "Don't boot unless everything loads." );
	}
}
```

or report in the DOM the failures

```ts
// Announce the failure, but keep on going.
onRunningSteps = steps => {
	steps.forEach( ( promise, step ) => {
		promise.catch( () => {
			const alert = document.createElement( 'div' );
			alert.classList.add( 'alert' );
			alert.innerHTML = `Failed to load '${ step.step }'`.
			document.querySelector( 'some selector' ).appendChild( alert );
		} );
	} );
}
```

obviously each of these could be inspected for better reporting or logic. in fact, maybe `onRunningSteps` isn't even as useful as `onStepsCompleted`.

eliot-akira replied:
> Is there an existing bug here where we should be awaiting the response from .resolve()? It looks like we may continue to lines 130-144 below before all the resources have resolved.

It's an unusual way to use Promises, to instantiate all of them without await, then, or catch. I had to think about this part when I read it first, but I understand it's intentional in this case, to start downloading resources from all steps in parallel, before executing any step. Internally, `resource.resolve()` uses a semaphore to manage a queue with a maximum number of concurrency.

> This may be correct, as we want to queue up all the async resources to start downloading in parallel, but we don't wait for the batch of them to complete.

A possible buggy part of this logic was that there was no `catch()` when a resource promise threw an error. It would show up in the browser dev console as an uncaught promise. With the change in this pull request, such errors are caught and passed to the `onStepError` callback, if any, or rethrown as an error of the async `CompiledBlueprint.run()` function for the caller to handle.

---

> I guess in `resolveArguments` we end up explicitly awaiting the results so that comes through. hm. this could use a comment at a minimum.

OK, I'll add:

```ts
/**
 * When each step's `CompiledStep.run()` is called, it uses
 * `resolveArguments` to await all dependent resources for the
 * step.
 */
```

Do you think the wording is clear enough?

---

> It seems that the onStepError callback might execute in the middle of processing or waiting for other resources. that also seems like it's probably fine, but possibly surprising.

That's true, the `onStepError` callback is called when `resource.resolve()` fails, or when `CompiledStep.run()` fails.

But in practice I think it's OK, because the callback receives the error as well as the step wherein it occurred, and a resource fetch can be considered part of the step: if it fails, the step fails, and the callback can decide if that's a show-stopper or to let the site load.

---

> - I don't care if one plugin fails to load as long as I can know it failed. keep going.
> - if my one particular thing fails then there's no point continuing.

Hmm, I lean toward the latter approach, I'd prefer to stop the whole thing if a plugin that I specified fails to load. Having a site in partially incomplete state seems like it would lead to unexpected errors down the road, for example in other steps that rely on a non-existent plugin.

That's how it worked originally, though without any error message to the user, I think, or it flashed an error in the progress screen, then continued loading the site without the plugin.

With the current state of this PR, there's no option in the Query API to specify that a plugin is *required* instead of optional, and to prevent loading the site if a plugin install fails.

In other words, I think we may need the opposite of the feature that the user was suggesting in #600, [An invalid plugin slug should not cause following plugins to not be installed](https://github.com/WordPress/wordpress-playground/issues/600). It's probably desirable in some situations, for an invalid plugin slug or plugin install error to stop the whole thing, rather than continue with steps that require it.

> in the end, I guess there's no harm in doing a partial boot of the playground as long as we're not connected to an OPFS or Live Directory share. partial boot in those places could corrupt files, though I hope people aren't relying on this not corrupting files. it's supposed to be ephemeral.

If a plugin is considered optional ("Let the site load even if it fails to install"), then I suppose it should be fine even with a persistent site. Who knows, the next time the plugin may successfully load! (Which is an approach I'm not that comfortable with, but apparently in some situations that is the desired behavior.)

---

> did you consider passing a callback when we finish running the steps? the current callbacks speak to individual steps, but not to the overall progress.

I hadn't considered it, but at first glance it seems like `onStepsCompleted` could simply be when the function returns? Or when it throw an error, instead of an `onStepsError` callback.

> Implementors could still replicate the behavior that `onStepCompleted` and `onStepError` provide, but they could additional prevent the boot-up of the playground.

Oh I see, you mean a different solution: instead of these step-specific callbacks, to have a single callback for all steps.

Hmm, but that approach would change the default behavior of throwing an error when any step fails (including steps to install plugins)?

I think the existing behavior ("Stop everything if a step or plugin install fails") is more intuitive, and that a fail-able step or an optional plugin is the exception, better suited as an opt-in behavior?

> obviously each of these could be inspected for better reporting or logic. in fact, maybe onRunningSteps isn't even as useful as onStepsCompleted.

How about, instead of an `onStepsCompleted` callback, the function `CompiledBlueprint.run()` could `return await Promise.allSettled( steps.values() )`? That could include return values of each step on success, or any error per step.

Well, I imagine it's do-able either way, to let the caller handle each error per step, or all step errors per blueprint. For example, the `onStepError` callback could display an alert and let the remaining steps continue or not.

eliot-akira replied:
> we may need the opposite of the feature that the user was suggesting in #600. It's probably desirable in some situations, for an invalid plugin slug or plugin install error to stop the whole thing, rather than continue with steps that require it.

Outside of the Query API, I think it makes sense for `CompiledBlueprint.run()` to throw if any step or plugin install fails, as default behavior, and optionally let a step be fail-able.

dmsnell replied:
> It's an unusual way to use Promises

it's not that unusual, especially in situations like this, maybe just less common than when working in async flows where we're waiting on individual promises in a chain of execution.

> onStepsCompleted could simply be when the function returns?

I like this idea, as long as we can signal back whether execution should proceed.

> Hmm, I lean toward the latter approach, I'd prefer to stop the whole thing if a plugin that I specified fails to load. Having a site in partially incomplete state seems like it would lead to unexpected errors down the road, for example in other steps that rely on a non-existent plugin.

here I wasn't tossing around two options we should choose from, but trying to list legitimate cases people will want. I don't think either one is a right choice but rather I think there will be plenty of situations where someone wants to make that decision themself, and might even make both decisions in a single playground. for example, _this_ plugin is a show-stopped but the theme is something I don't care about, or the caching plugin.

it's more about giving control to people building sites or tools with the Playground. the idea is that if we can report the progress or completion status then they can indicate whether to boot up the Playground or terminate with an error.

> that approach would change the default behavior of throwing an error when any step fails (including steps to install plugins)?

yeah, and I wouldn't think it's as important if we signal this at every step as it fails or if we wait until the end and list all the completion statuses. wouldn't it only change the behavior from what's proposed already in this PR, though, not how it works today? (since today it doesn't report any errors)

> If a plugin is considered optional ("Let the site load even if it fails to install"), then I suppose it should be fine even with a persistent site. Who knows, the next time the plugin may successfully load! (Which is an approach I'm not that comfortable with, but apparently in some situations that is the desired behavior.)

right, and that's why I think a few people have reported back that they want the Playground to continue booting when something doesn't load. the important bits are there, the extras aren't.

if the right steps are required then it's fine on a persisted site, but if, for example, some plugin transforms data on save or maintains table relationships, and it doesn't load for a persistent site, then we could get into situations where it corrupts the database or file uploads just because the site continues to run without part of its functionality. in this case that would be caused by someone not instructing the thing to fail on error, but it would be something we would want to make sure we communicate could happen.

I only wanted to mention that there's this difference in risk for temporary sites vs. those persisting their data, particularly those using the Live Directory in Chrome or using `wp-now` or otherwise using the Playground while connected to real data.

eliot-akira replied:
I can try to rewrite the logic and implement `onStepsCompleted` instead of `onStepError` and `onStepCompleted` (which was already there before this PR) - but I'm not sure I understand the proposed change.

```ts
	for (const { run, step } of compiled) {
		try {
			const result = await run(playground);
			progressPromises.get(step).resolve();
		} catch (error) {
			progressPromises.get(step).reject();
		}
	}
} finally {
	try {
		if (onStepsCompleted) {
			await onStepsCompleted( steps );
		}
```

Within the catch clause for each step, if it only sets the rejected status for each promise, it seems there isn't an opportunity for the caller of `CompiledBlueprint.run()` (such as the website or on server side) to abort the process, unlike with the `onStepError` callback? And a failed step would always let remaining steps continue?

..Ah I see, that's the purpose of the `onRunningSteps` callback. It loops through the step promises beforehand, so it can attach `catch()` before starting the execution of steps. And if there is no `onRunningSteps` callback passed, then a failed step promise will let other steps continue. Such failed steps are passed to `onStepsCompleted()`, which can choose to continue loading the site (and call `progress.finish()`), or stop and show an error.

OK, I think I understand the intent. I can rewrite the logic in this way.

One point I'm not so convinced about, that the default behavior of a failed step should be to continue running the rest of the steps. Unlike on the Playground website, when using the Blueprints API from the Playground Client module or from Node.js like with `wp-now`, I think it's more intuitive for a step error to be a show-stopper by default?

eliot-akira replied:
> I only wanted to mention that there's this difference in risk for temporary sites vs. those persisting their data, particularly those using the Live Directory in Chrome or using wp-now or otherwise using the Playground while connected to real data.

Right, I see what you mean about the risk of corrupting data. This reenforces my hesitation to let a site continue loading in partial/incomplete state, i.e., when a required plugin is not installed. I think it should be opt-in behavior.

---

Just as an idea, if we decide to allow plugin installs to fail, in other words they're optional by default for the Query API - maybe there could be a syntax to append or prepend `!` (exclamation mark) to the plugin slug to mark it as required.

(..Meh, on second thought, that's not that great, it will be another thing people would have to learn. The Query API is already a bit complex to use, like the syntax of installing multiple plugins with pairs of `plugin=value` instead of a comma-separated list. In that case too, I remember the idea of a "micro-syntax" was declined due to potentially introducing complexity by parsing plugin URLs. Never mind.)

eliot-akira replied:
Here's how it looks with the logic changed from using "per-step callbacks", `onStepCompleted` and `onStepError`, to using "per-blueprint callbacks", `onRunningSteps` and `onStepsCompleted`.

```ts
const progressMap = new Map<StepDefinition, Promise<void>>();
const progressPromises = new Map<StepDefinition, {resolve: () => void, reject: (error: string) => void}>();
for (const { step } of compiled) {
	progressMap.set(step, new Promise((resolve, reject) => {
		progressPromises.set(step, {resolve, reject});
	} ) );
}
onRunningSteps( progressMap );

try {
	// Start resolving resources early
	for (const { resources, step } of compiled) {
		for (const resource of resources) {
			resource.setPlayground(playground);
			if (resource.isAsync) {
				resource.resolve().catch(function (error) {
					// onStepError(error, step);
					progressPromises.get(step).reject(error);
				});
			}
		}
	}

	/**
	 * When each compiled step's `run` method is called, it uses
	 * `resolveArguments` to await dependent resources for the step.
	 */
	for (const { run, step } of compiled) {
		try {
			const result = await run(playground);
			// onStepCompleted(result, step);
			progressPromises.get(step).resolve(result);
		} catch (error) {
			// onStepError(error, step);
			progressPromises.get(step).reject(error);
		}
	}
} finally {
	try {
		if (onStepsCompleted) {
			await onStepsCompleted( progressMap );
		}
		await (playground as any).goTo(
			blueprint.landingPage || '/'
		);
	} catch (e) {
		/*
			* NodePHP exposes no goTo method.
			* We can't use `goto` in playground here,
			* because it may be a Comlink proxy object
			* with no such method.
			*/
	}
	progress.finish();
}
```

---

The new logic will accomplish basically the same thing as per-step callbacks, namely to provide the caller an ability to interrupt or continue running the remaining steps when a step fails.

A possible disadvantage is that it's a bit more complicated for the caller. They must know how the `onRunningSteps` callback is meant to work, by looping through promises to attach `catch` (and maybe `then`?) callbacks. I think per-step callbacks are easier to understand and use, by being passed the step result or error directly.

One advantage of the per-blueprint callbacks that I can see, is that in addition to the ability to interrupt/continue steps after a step fails, it gives the ability after all steps have finished running, to decide whether to continue loading the site with `playground.goTo( landingPage )` based on the step results. (But the `onStepError` callback can do that too, just not at the end but when an error occurs.)

Another advantage is that the `onStepsCompleted` callback gets the results (including any errors) of all steps. But then again, the per-step callbacks get these too, just not at the end but as each step result/error occurs.

eliot-akira replied:
> right, and that's why I think a few people have reported back that they want the Playground to continue booting when something doesn't load. the important bits are there, the extras aren't.

So, with the current state of this PR being "Continue plugin installs on error", we're letting the remaining steps continue and the site to load when *any* plugin fails to install. In effect, for the website Query API, we would be considering all plugins to be optional ("extras") and not strictly required ("important"). I wonder if that's what we want.

For the Query API, I think it does make sense in terms of user experience, to let the website load even if a plugin install fails. That's far better than getting an error screen and being unable to play with the site at all, even if it's partial/incomplete.

For all other uses of the Blueprint module, via Playground Client or on the server side, I think it's more intuitive and common practice to throw any error as soon as it occurs; and only optionally let the caller decide if a step is optional and fail-able.

dmsnell replied:
> One point I'm not so convinced about, that the default behavior of a failed step should be to continue running the rest of the steps…it's more intuitive for a step error to be a show-stopper by default?

that's fine because I'm not trying to convince of that 😄 I just want to make it possible for someone to determine that, and right now without your awesome contribution here it's just not possible. so I'm not arguing to keep going; I'm only arguing to build this system which lets _other people_ choose, and specifically avoiding _us_ making that call for them.

> maybe there could be a syntax to append or prepend ! (exclamation mark) to the plugin slug to mark it as required.
> (..Meh, on second thought, that's not that great, it will be another thing people would have to learn. The Query API is already a bit complex to use

I concur with you here and I think that if someone wants something more complicated they can already do that. in follow-up work I think we could look at the plugin and theme blueprint steps and add an `optional` flag to indicate that it's fine to continue if they fail. this would be the easy canonical way for people to set that up, and it would be a short jump from calling `?plugin=my-plugin` to adding `#{"steps":[{"step":"installPlugin", "pluginName": "my-plugin"}]}` (yes it's verbose and yes I think there is another step involved here in getting the plugin zip, but these things can also be handled through a better UI and documentation - a blueprint fragment builder of sorts, exposing `compileBlueprints` yada yada).

---

concerning the callback maybe we have some iteration to do on the design.

> to abort the process, unlike with the onStepError callback? And a failed step would always let remaining steps continue?

okay I think I am conflating two things: steps and resources. resources can load concurrently but steps are sequenced.

```js
// Start resolving resources early
for (const { resources, step } of compiled) {
	for (const resource of resources) {
		resource.setPlayground(playground);
		if (resource.isAsync) {
			resource.resolve().catch(function (error) {
				// onStepError(error, step);
				progressPromises.get(step).reject(error);
			});
		}
	}
}
```

in this chunk we start fetching all resources concurrently. we have no idea in what order they are going to resolve, so the `onStepError` callback at this point doesn't provide much more than a single callback after all resources are resolved. yes, we may call it as soon as the first dependent resource fails for a given step, but that could also be the last of many resources to fetch. we can imagine such a scenario when we're installing a plugin and a theme. even if the plugin is the first step, the server could be slow sending the ZIP and the theme resource might appear first. in this case, by the time the plugin "fails" the theme resources are already here.

<img src="https://japari-library.com/w/images/0/0c/WhackAMoleUniqueBehavior.gif" align=right>

this is why I first thought of the promise map, because I imagined something like a whack-a-mole board where steps settle as their dependent resources fully succeed or fail. this is also why I was first confused by `resolveArguments` because we really don't know after this nested for loop if a step is ready, we have to wait for `resolveArguments` (or check the status of each promise in the resources for that step). that could involve more than fetches too, as dependent directory creation or file copying could also fail.

though I didn't show it in my code snippet, I was thinking for the `onStepsCompleted` callback, something like this would work

```js
if ( onStepsCompleted && await onStepsCompleted( progressMap ) ) {
	await playground.goTo( blueprint.landingPage || '/' );
} else {
	showErrorScreen();
}
```

we don't have to expose the promise map at this point either: we could reduce it to a boolean status of step -> status

first, a tiny helper that I'll use a couple of times here

```ts
/**
 * Convert a Promise into a boolean indicating if it resolved or rejeted.
 */
const didResolve = <P extends Promise<any>>(promise: P): bool =>
	p.then( () => true, () => false );
```

```js
const completedStatus = new Map();
for ( const { step } of compiled ) {
	completedStatus.set(
		step,
		await didResolve( progressMap.get( step ) )
	);
}
onStepsCompleted( completedStatus );
```

my earlier ideas about `onRunningSteps` was that it would give the caller more control to totally abort everything once something fails, as in, to stop downloading a large dependent ZIP file.

```js
onRunningSteps = ( steps, { abortController } ) => {
	// Without my plugin, nothing should boot up.
	steps
		.find( ( { step, pluginName } ) => step === 'activatePlugin' && pluginName === 'my-important-plugin' )
		.catch( () => abortController.abort() );
}
```

maybe that's more of an enhancement than a necessary part of this. it all still gives a place to error out immediately once a step fails, but only if someone wants that. so maybe the default should finally look something like this:

```js
// Either someone made a choice to boot or all steps succeeded.
const shouldBoot = onStepsCompleted
	? await onStepsCompleted( completedStatus )
	: await didResolve( Promise.all( [ ...progressMap.values ] ) );

if ( shouldBoot ) {
	await playground.goTo( blueprint.landingPage || '/' );
} else {
	showErrorScreen();
}
```

with this, all someone needs to do is write a function that examines the steps and a `true` or `false` value if they succeeded, and return `true` or `false` to indicate if the Playground should proceed.

if we add `optional` to plugin and theme Blueprint steps, all we have to do is implement this ourself with a custom `onStepsCompleted` handler.

thanks for going through all this discussion and back and forth. I think it's good that we struggle with it now rather than after we publish and people start building on it.

eliot-akira replied:
> in follow-up work I think we could look at the plugin and theme blueprint steps and add an optional flag to indicate that it's fine to continue if they fail.

I created a new issue for it.

- #621

---

> a blueprint fragment builder

That's an idea, a visual interface to generate a URL with blueprint steps encoded as query parameters. I captured it in a new issue.

- #622

---

> we start fetching all resources concurrently. we have no idea in what order they are going to resolve, so the onStepError callback at this point doesn't provide much more than a single callback after all resources are resolved.

Very true. That part where it calls the `onStepError` callback on resource resolve fail, I didn't have a good feeling about it because it's before and outside of the loop where steps are run sequentially. As you pointed out, resources are fetched in parallel, so a step can fail in any order in whack-a-mole fashion.

Right, I see how a promise map can solve this situation by deferring acting on the resolve/reject status of a resource until its step is being run. I like that solution.

OK, I'll implement `onRunningSteps` and `onStepsCompleted` as you suggested, I think I understand how all the parts fit together.

> it's good that we struggle with it now rather than after we publish

Agreed, thank you for the thoughtful review and feedback. I'll keep tenacious and persistent as we refine the details.

eliot-akira replied:
- Commit 1281355: [Implement `onRunningSteps` for success/error handling per step; and `onStepsCompleted` for success/error handling per blueprint](https://github.com/WordPress/wordpress-playground/commit/1281355bbb992e74245a55e581ed1249d71e20c2)

dmsnell replied:
noice. I'm going to mark this thread resolved because it's getting in the way with its length. feel free to unresolve it, or open a new thread to continue chatting about it.

dmsnell commented on packages/playground/client/src/index.ts:
```diff
@@ -47,7 +48,8 @@ export interface StartPlaygroundOptions {
 	blueprint?: Blueprint;
 	onBlueprintStepCompleted?: OnStepCompleted;
-	onBlueprintStepError?: OnStepError;
+	onBlueprintStepsCompleted?: OnStepsCompleted;
```
now that I look at these side by side it feels odd that they are structured differently. do you have ideas for names that might be more symmetric?

```
onBlueprintStepCompleted
onBlueprintStepsCompleted
onBlueprintStepsRunning
```

```
onBlueprintStepsStarting
onBlueprintStepsProgress
onBlueprintStepsFinished
```

dmsnell commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -48,15 +48,36 @@ export interface CompiledBlueprint {
+export type PromiseExecutor = {
+	resolve: (result: any) => void;
+	reject: (error: any) => void;
+};
```
I don't particularly find these custom types that much more helpful than the concrete types they alias. e.g.

```ts
export type OnStepsCompleted = (completedStatus: Map<StepDefinition, boolean>) => void;
export type OnRunningSteps = (progressMap: Map<StepDefinition, Promise<void>) => void;
```

this is succinct and I don't have to make sure the named types are hiding anything or doing anything surprising.

I guess you found otherwise? what benefit do you see from having them named? you think it's a worthwhile tradeoff?

dmsnell replied:
as with the naming, I think there's room for harmony in the name of the argument. I like `completedStatus` because it indicates what it means, but even though I originally wrote `progressMap` I find it a bit low-level and internal. maybe `progressStatus`?

eliot-akira replied:
I consolidated the changes into a single `onStepProgress` callback, similar in concept to `onStepCompleted`.

dmsnell commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -71,6 +92,14 @@ export function compileBlueprint(
+			for (const promise of progressMap.values()) {
+				promise.catch((error) => {
+					throw error;
+				});
```
is this necessary? won't it already raise an error if we don't handle the catch in an `async`/`await` block?

eliot-akira replied:
Removed.

dmsnell commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -116,33 +145,77 @@ export function compileBlueprint(
+				for (const { resources, step } of compiled) {
+					const stepPromise = progressPromises.get(
+						step
+					) as PromiseExecutor;
```
this `PromiseExecutor` shouldn't be necessary. was it telling you it could be `undefined`? If that's the case, we might be able to try `progressPromises.get(step)!`

dmsnell commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -116,33 +145,77 @@ export function compileBlueprint(
+						stepPromise.resolve(result);
+						onStepCompleted(result, step);
+					} catch (error) {
+						stepPromise.reject(error);
```
we could also call these something else to explain more the use.
here's one random idea:

```js
const stepIndicator = stepIndicators.get( step );

…

stepIndicator.succeedWith( result );

…

stepIndicator.failWith( error );
```

or `succeededWith`/`failedWith`

```js
stepIndicators.set( step, { succeedWith: resolve, failWith: reject } );
```

eliot-akira replied:
`stepPromise` has been renamed to `stepStatus`.

dmsnell commented on packages/playground/website/src/lib/hooks.ts:
```diff
@@ -36,10 +37,23 @@ export function usePlayground({ blueprint, storage }: UsePlaygroundOptions) {
 		}
 
+		const onRunningBlueprintSteps = (progressMap: PromiseMap) => {
+			for (const [step, promise] of progressMap.entries()) {
```
this is a good place to look at how the naming of the values affects the readability of the intent in the code

eliot-akira replied:
Refactored as `onBlueprintStepProgress(promise, step)`.

dmsnell reviewed (commented):
love how this is shaping up. makes me want to see if we can push forward on the toast component

adamziel commented:
@eliot-akira I'd love to have a mechanism of handling errors (and output) baked into Blueprints and it's becoming increasingly relevant for the extenders, too. Would you be open to doing a few more iterations here? Or, if you don't have the bandwidth at the time, would you mind if I pushed some commits this PR?

eliot-akira commented:
> would you mind if I pushed some commits this PR?

@adamziel Sure, please do. To be honest I don't think I have the bandwidth to carry this through to the finish line in a timely manner. There are some unresolved questions about the specifics of how the feature works internally, and also the public interface like the naming of options/methods, and how users are expected to use this feature. It probably will require making detailed decisions toward a clear concept/design.

@dmsnell Sorry I dropped the ball on this one, I got swept away by other concerns and haven't had the time or mental space to discuss and solve this completely. Looks like I bit off more than I could chew!

eliot-akira commented:
OK, I refactored the changes in this PR into a new option `onStepProgress()`, a callback that receives the step that's running and a promise in progress, which can be awaited or caught for error handling. I also reduced the number of types being added to a minimum, by inlining those that are short or used only once.

I haven't tested yet if this works as expected, in particular:

- If there is no `onStepProgress` callback passed, it should keep existing behavior by throwing an error for any failing step (such as plugin install), and stop the whole blueprint
- The `onStepProgress` callback should be able to:
  - Catch any error to let the rest of blueprint steps continue
  - Throw an error and stop the whole blueprint
 
Also, a question remains if this interface is designed in a way that can support the implementation of optional or fail-able steps, such as a step option that indicates whether the step is required or not for the rest of the blueprint.

So the public interface is still a draft in progress, to see how best to solve these related issues.

- WordPress/blueprints#45
- WordPress/blueprints#40
- WordPress/wordpress-playground#600
- WordPress/wordpress-playground#604

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -48,17 +48,35 @@ export interface CompiledBlueprint {
+/**
+ * Convert a Promise into a boolean indicating if it resolved or rejeted.
+ */
+const didResolve = <P extends Promise<any>>(promise: P): Promise<boolean> =>
```
`didResolve() === false` suggests the promise may still be pending. I wonder what would be a more descriptive name here.

eliot-akira replied:
It was tough to find a suitable naming for this `didResolve` function. I consulted [MDN: Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) to see what the standard terminology is.

> A Promise is in one of these states:
>
> - **pending**: initial state, neither fulfilled nor rejected.
> - **fulfilled**: meaning that the operation was completed successfully.
> - **rejected**: meaning that the operation failed.
>
> A promise is said to be **settled** if it is either fulfilled or rejected, but not pending.

In the end I settled on a verbose name `isSettledPromiseResolved`, to make it clear that it's checking an already settled promise for its resolved status.

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -171,33 +191,81 @@ export function compileBlueprint(
+					try {
+						const result = await run(playground);
+						stepPromise?.resolve(result);
+						onStepCompleted(result, step);
```
If the `onStepCompleted` handler throws an error, the `stepPromise?.reject(error);` will run. It would be nice to just rethrow that error so the API consumer may notice and correct it.

eliot-akira replied:
Instead of rethrowing, the two lines `stepStatus?.resolve(result)` and `onStepCompleted` have been moved below the try/catch clause. If the latter throws (by the user or an unexpected error), it bubbles up to the API consumer, the caller of `CompiledBlueprint.run()`.

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -171,33 +191,81 @@ export function compileBlueprint(
 		},
 		run: async (playground: UniversalPHP) => {
+			const progressStatus = new Map<StepDefinition, Promise<void>>();
+			const progressPromises = new Map<
```
`progress` as a word is already used for tracking the percentage 0 - 100% completion status. How about `stepPromises`?

eliot-akira replied:
They've been renamed to `stepPromises` and `stepStatus`.

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -171,33 +191,81 @@ export function compileBlueprint(
 			networking: blueprint.features?.networking ?? false,
 		},
 		run: async (playground: UniversalPHP) => {
+			const progressStatus = new Map<StepDefinition, Promise<void>>();
```
Same not as below – how about `stepStatus`?

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -48,17 +48,35 @@ export interface CompiledBlueprint {
 	/** Optional callback with step output */
+	onStepProgress?: OnStepProgress;
 	onStepCompleted?: OnStepCompleted;
+	onStepsCompleted?: OnStepsCompleted;
```
The intention seems to be deciding whether to boot. How about `shouldBoot` as a name?

eliot-akira replied:
`onStepsCompleted` has been renamed to `shouldBoot`.

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -171,33 +191,81 @@ export function compileBlueprint(
+						stepPromise?.resolve(result);
+						onStepCompleted(result, step);
+					} catch (error) {
+						stepPromise?.reject(error);
```
I wonder whether the failure consequences should be possible to specify per step type, e.g. `rm` could default to triggering a warning but not stopping the execution, while `defineWpConfigConsts` would default to throwing an error and stopping the execution.

There seem to be a hierarchy of control here:

1. Consumer has a preference. If missing, fall back to...
2. Step has a preference. If missing, fall back to...
3. Blueprint engine has a preference – that's implicit from the code structure here

It would be nice to stop the execution as soon as we know the failure is fatal. Imagine the first step of a work-intense Blueprint fails and we already know this is irrecoverable – stopping right then and there would spare the user 30s of waiting before learning about the error.

Code-wise, it could mean a logic similar to this:

```suggestion
						stepPromise?.reject(error);
						if( isFailureFatal( step, error ) ) {
							throw e;
						}
```

Now, the `isFailureFatal` would need to know about both the step preference and the user preference, which means we need to provide developers with a way of overriding the default behavior on a granular, per-step basis.

This PR introduces a decision point in the `onStepsCompleted` callback registered by the application calling the `compileBlueprint` function. I think it would be more useful to make that decision inside of a Blueprint. Imagine a live product demo. The `installPlugin` installing the demoed plugin is indispensable, but the next one installing a dev utility plugin can likely be skipped if it fails.

Technically, this could mean introducing a new step-level property:

```diff
export type StepDefinition = Step & {
	progress?: {
		weight?: number;
		caption?: string;
	};
+	onFailure: 'abort' | 'skip';
};
```

This would enable writing blueprints such as...

```json
{
	"steps": [
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "my-product"
			},
			"onFailure": "abort"
		},
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "devtools"
			},
			"onFailure": "skip"
		}
}
```

The default value could be `abort`. Each step handler would be able to override it, for example like this:

```ts
/**
 * Removes a file at the specified path.
 */
export const rm: StepHandler<RmStep> = async (playground, { path }) => {
	await playground.unlink(path);
};

import { defaultStepFailureHandlers } from '.';
defaultStepFailureHandlers.set( 'rm', 'continue' );
```

The `compileStep()` function could then output the desired failure behavior:

```ts

function compileStep<S extends StepDefinition>(
	step: S,
	{
		semaphore,
		rootProgressTracker,
		totalProgressWeight,
	}: CompileStepArgsOptions
): { run: CompiledStep; step: S; resources: Array<Resource> } {
	return {
		run,
		step,
		resources,
		onFailure:  'onFailure' in step ? step.onFailure : defaultStepFailureHandlers.get( step.step ) ?? 'abort'
	};
```

And this loop could handle it as follows:

```ts
				for (const { run, step, onFailure } of compiled) {
					const stepPromise = progressPromises.get(step);
					try {
						const result = await run();
						onStepSettled(step, SUCCESS, result);
					} catch(e) {
						onStepSettled(step, FAILURE, e);
						if( onFailure === 'abort' ) {
							throw e;
						}
					}
```

adamziel replied:
I wonder how the ideas above resonate with you @eliot-akira @dmsnell @bgrgicak

dmsnell replied:
I think it fits within the idea I had, which was to expose each to the application and let it decide.

we could provide default behaviors, which I think should be rare, and then if someone in userspace adds their own handler for the step, then they could take their own action.

or, we could expose our default handling to the callback and only if they implement the callback would we use theirs instead.

bgrgicak replied:
I like this approach, it would at least help with development and while creating blueprints.

Personally, I love opinionated code that I can adjust, so I would go with having defaults for each step, but let users override them. Maybe even include a global override for all steps.

eliot-akira replied:
Thank you for the detailed review and feedback. I'm starting to get a feel for the public interface that's shaping up.

I like the idea of adding to the `StepDefinition` a new property `onFailure`, with value `abort` or `skip`. That addresses the original issue that motivated this PR, which is to allow some plugins to be skipped if install fails, while still keeping plugins required by default (error on install fail).

It makes sense for each step handler to have a default failure mode (`abort` or `skip`), which can be overridden by the user.

I'll get started on implementing the suggestions, and will either resolve each reviewed point or reply with any questions.

adamziel replied:
Thank you so much @eliot-akira, you are the best!

eliot-akira commented:
OK, I implemented the suggestions with my own take on some of the details.

I've added a blueprint step option called `failMode`, with possible value `abort` or `skip`. The naming seemed more suitable compared to `onFailure`, `onFail`, or `onError` which sounds like a callback function.

It meets these requirements, which I wrote down before starting this round of changes.

- [x] Skip mode: Catch any error to let the rest of blueprint steps continue
- [x] Abort mode: Throw an error and stop the whole blueprint
  - [x] Stop the execution as soon as we know the failure is fatal

---

The compileBlueprint option `onStepsCompleted` has been renamed to `shouldBoot`, which takes a map of step completion status, and returns a boolean to decide if the site should continue booting.

I also added a compileBlueprint option `onStepError`, to let the website have a chance to display an error message in a modal - currently using `alert()`.

The callback can optionally throw an error to abort the rest of the steps (same as `abort` mode), or let them continue (same as `skip` mode). This part I'm a bit unsure of the logic, how it interacts with the fail mode.

---

It's going in the right direction I think, but would appreciate another review.

For manual testing:

- This URL should show an error message, then continue booting the site: http://localhost:5400/website-server/?plugin=not-existing

eliot-akira commented on packages/playground/website/src/lib/hooks.ts:
```diff
@@ -40,10 +41,25 @@ export function usePlayground({ blueprint, storage }: UsePlaygroundOptions) {
+				// TODO: Open modal to display error message
+				alert(
+					`Failed to install: ${
+						(step.pluginZipFile as CorePluginReference).slug ||
```
This type coersion to `CorePluginReference` is questionable. It was a temporary workaround because `step.pluginZipFile` is a `FileReference`, a union of several types which have property `slug`, `name`, `path`, or nothing for the name.

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -171,33 +195,92 @@ export function compileBlueprint(
+				for (const { step } of compiled) {
+					completedStatus.set(
+						step,
+						await isSettledPromiseResolved(
```
Will this wait forever if the `finally` section is triggered by an error thrown before all the steps are run?

adamziel commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -171,33 +195,92 @@ export function compileBlueprint(
+				// Either someone made a choice to boot or all steps succeeded.
+				const boot = shouldBoot
+					? shouldBoot(completedStatus)
+					: await isSettledPromiseResolved(
```
At this point, no further steps are executed. Also, nothing seems to await `stepPromises` – do we need to reason about status as promises, then? Or could the status information be stored in a synchronous data structure like `Map<StepDefinition, { didRun, error, result }>`?

adamziel commented:
There's some really good ideas in this PR. Since Blueprints are transitioning into a PHP library, let's freeze the set of features offered by the TypeScript implementation and keep exploring those ideas in the new, dedicated repository: https://github.com/WordPress/blueprints/. I'll go ahead and close this PR. Thank you so much for your work @eliot-akira!

eliot-akira commented:
> Blueprints are transitioning into a PHP library, let's freeze the set of features offered by the TypeScript implementation

I was wondering how that effort related to this PR. That makes sense to focus on the new implementation. I'll follow along with the progress there.

---

To be honest I was struggling a bit with this PR, and still wasn't totally clear on how the async/promises logic was supposed to work.

Off topic, but recently I had a somewhat similar situation - start many concurrent tasks (convert incoming text to speech) while handling their results sequentially (play the generated audio) - and I realized I need more practice to be fluent with async stuff.

In that case, I discovered what would really help solve it was [`Promise.withResolvers()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers).

> The key difference when using `Promise.withResolvers()` is that the resolution and rejection functions now live in the same scope as the promise itself, instead of being created and used once within the executor.
>
> This may enable some more advanced use cases, such as when reusing them for recurring events, **particularly with streams and queues**. This also generally results in less nesting than wrapping a lot of logic within the executor.

It probably would have helped simplify some of the logic in this PR, but it's not fully supported by JavaScript runtimes yet (notably Safari/WebKit and Node.js).
